README.md
# Tookit

Core utility library for Clojure.

## Usage

FIXME: write usage documentation!

Invoke a library API function from the command-line:

    $ clojure -X com.bobnadler.toolkit/foo :a 1 :b '"two"'
    {:a 1, :b "two"} "Hello, World!"

Run the project's tests (they'll fail until you edit them):

    $ clojure -T:build test

Run the project's CI pipeline and build a JAR (this will fail until you edit the tests to pass):

    $ clojure -T:build ci

This will produce an updated `pom.xml` file with synchronized dependencies inside the `META-INF`
directory inside `target/classes` and the JAR in `target`. You can update the version (and SCM tag)
information in generated `pom.xml` by updating `build.clj`.

Install it locally (requires the `ci` task be run first):

    $ clojure -T:build install

Deploy it to Clojars -- needs `CLOJARS_USERNAME` and `CLOJARS_PASSWORD` environment
variables (requires the `ci` task be run first):

    $ clojure -T:build deploy

Your library will be deployed to com.bobnadler/toolkit on clojars.org by default.

## License

Copyright © 2022 Bob Nadler, Jr.

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Removed
### Fixed

[Unreleased]: https://github.com/bobnadler/toolkit


test/com/bobnadler/toolkit_test.clj
(ns com.bobnadler.toolkit-test
  (:require [clojure.test :as t]
            [com.bobnadler.toolkit :as bntk]))

(t/deftest redact-map-test
  (let [options {:redact-key? #{:secret :password} :redact-value "[FILTERED]"}]
    (t/is (= {:secret "[FILTERED]"}
             (bntk/redact-map {:secret "123"} options)))

    (t/is (= {:secret "[FILTERED]" :foo 42}
             (bntk/redact-map {:secret "123" :foo 42} options)))

    (t/is (= {"secret" "[FILTERED]" :foo 42}
             (bntk/redact-map {"secret" "123" :foo 42} options)))

    (t/is (= {:nested {"secret" "[FILTERED]"} :foo 42}
             (bntk/redact-map {:nested {"secret" "123"} :foo 42} options)))

    (t/is (= {:nested {"secret" "[FILTERED]"} :password "[FILTERED]"}
             (bntk/redact-map {:nested {"secret" "123"} :password "secret"} options)))))

(t/deftest truncate-start-test
  (let [long-str "A long string that should be truncated"
        short-str "A short string"]
    (t/testing "with defaults"
      (t/is (= "...at should be truncated"
               (bntk/truncate-start long-str)))
      (t/is (= "A short string"
               (bntk/truncate-start short-str))))

    (t/testing "alternate prefix"
      (t/is (= "$$$at should be truncated"
               (bntk/truncate-start long-str "$$$" 25)))
      (t/is (= "A short string"
               (bntk/truncate-start short-str "$$$" 25))))

    (t/testing "alternate limit"
      (t/is (= "...ould be truncated"
               (bntk/truncate-start long-str "..." 20)))
      (t/is (= "A short string"
               (bntk/truncate-start short-str "..." 20))))))

(t/deftest wrap-request-id-test
  (let [handler (bntk/wrap-request-id identity)]

    (t/testing "no header"
      (t/is (contains? (handler {}) :request-id)))

    (t/testing "header is present"
      (let [req-id "c9bd7379-3fe9-45e8-8f69-a8e20fa007a7"]
        (t/is (= req-id
                 (:request-id (handler {:headers {"x-request-id" req-id}}))))))))

(t/deftest wrap-request-start-logger-test
  (let [state_ (atom nil)
        fake-logger (fn [log-info] (reset! state_ log-info))
        req {:request-method :get :uri "/foo" :query-string "bar=2"}
        handler (bntk/wrap-request-start-logger identity {:log-fn fake-logger})]

    (t/testing "message is logged"
      (handler req)
      (t/is (= {:level :info
                :msg "Started GET '/foo?bar=2'"
                :attrs {:method "GET" :path "/foo?bar=2"}}
               @state_)))

    (t/testing "request map is returned"
      (let [response (handler req)]
        (t/is (= :get (:request-method response)))
        (t/is (= "/foo" (:uri response)))
        (t/is (= "bar=2" (:query-string response)))
        (t/is (> (:start-ms response) 0))))))

(t/deftest wrap-request-finish-logger-test
  (let [state_ (atom nil)
        fake-logger (fn [log-info] (reset! state_ log-info))
        req {:request-method :get :uri "/foo" :query-string "bar=2"}
        handler (bntk/wrap-request-finish-logger
                 #(assoc % :status 200)
                 {:log-fn fake-logger})]

    (t/testing "message is logged with duration"
      ;; Might be brittle, but works for now
      (handler (assoc req :start-ms (System/currentTimeMillis)))
      (let [{:keys [level msg attrs]} @state_]
        (t/is (= :info level))
        (t/is (= "Completed GET '/foo?bar=2' 200 in 0ms" msg))
        (t/is (= "GET" (:method attrs)))
        (t/is (= "/foo?bar=2" (:path attrs)))
        (t/is (= 200 (:status attrs)))
        (t/is (= "0ms" (:duration attrs)))))

    (t/testing "message is logged with unknown duration"
      (handler req)
      (let [{:keys [level msg attrs]} @state_]
        (t/is (= :info level))
        (t/is (= "Completed GET '/foo?bar=2' 200 in ?ms" msg))
        (t/is (= "GET" (:method attrs)))
        (t/is (= "/foo?bar=2" (:path attrs)))
        (t/is (= 200 (:status attrs)))
        (t/is (= "?ms" (:duration attrs)))))

    (t/testing "request map is returned"
      (t/is (= (assoc req :status 200) (handler req))))))

(t/deftest wrap-request-params-logger
  (let [state_ (atom nil)
        fake-logger (fn [log-info] (reset! state_ log-info))
        handler (bntk/wrap-request-params-logger identity {:log-fn fake-logger})
        request {:params {:foo 42 :password "secret"}}]
    (t/testing "message is logged with filtered params"
      (handler request)
      (let [{:keys [level msg attrs]} @state_]
        (t/is (= :info level))
        (t/is (= "Request parameters" msg))
        (t/is (= {:foo 42 :password "[FILTERED]"} (:params attrs)))))

    (t/testing "anti-forgery token is removed from params"
      (handler {:params {:foo 42 :__anti-forgery-token "abc123"}})
      (let [{:keys [attrs]} @state_]
        (t/is (= {:foo 42} (:params attrs)))))

    (t/testing "no message emitted if params are empty"
      (reset! state_ nil)
      (handler {})
      (t/is (nil? @state_)))

    (t/testing "request map is returned"
      (t/is (= request (handler request))))))


src/com/bobnadler/toolkit.clj
(ns com.bobnadler.toolkit
  "Various utility functions for Clojure projects."
  (:require
   clojure.set
   [clojure.string :as s]
   [clojure.walk :as walk]))

(defn redact-map
  "Returns a new map where the given map `m`'s values have been replaced by
  `redact-value` for any key that matches `redact-key?`."
  [m {:keys [redact-key? redact-value]}]
  (walk/postwalk
   (fn [x]
     (if (map? x)
       (->> x
            (map (fn [[k v]]
                   (if (redact-key? (keyword k))
                     [k redact-value]
                     [k v])))
            (into {}))
       x))
   m))

(comment
  (redact-map {:foo "bar" :password "secret"}
              {:redact-key? #{:password}
               :redact-value "[FILTERED]"})) ; {:foo "bar", :password "[FILTERED]"}

(defn truncate-start
  "Returns `s` truncated from the beginning instead of the end.

    prefix: String to prepend to the truncated target string.
    limit:  Number of characters to limit `s`. If the length of `s` is
            smaller than `limit`, do not truncate `s`."
  ([s prefix limit]
   (let [s-length (count s)
         target-length (- limit (count prefix))]
     (if (<= s-length target-length)
       s
       (str prefix (subs s (- s-length target-length) s-length)))))

  ([s] (truncate-start s "..." 25)))

(comment
  (def long-str "this is a long string that should be truncated")
  (truncate-start long-str) ; "... t should be truncated"
  (truncate-start "a short string") ; "a short string"
  (count (truncate-start long-str))) ; 25

(defn wrap-request-id
  "Ring middleware that wraps a request with an identifier.

  Checks the request for a `x-request-id` header, if one is present it is
  added to the request map. If not present one is generated and added to the
  request map.

  Options:

    log-context-fn: A fn that provides context for all log requests. For
                    example, when using Timbre...

                    (fn [ctx f] (timbre/with-context+ ctx (f)))"
  [handler & {:keys [log-context-fn]}]
  (fn [request]
    (let [request-id (get-in request [:headers "x-request-id"] (java.util.UUID/randomUUID))
          request-id-str (.toString request-id)]
      (if log-context-fn
        (log-context-fn
         {:request-id request-id-str}
         #(handler (assoc request :request-id request-id-str)))
        (handler (assoc request :request-id request-id-str))))))

(comment
  (def handler (wrap-request-id identity))

  (handler {}) ; {:request-id "2c8d4796-dc7a-47db-8182-f172b1f06521"}

  (handler {:headers {"x-request-id" "c9bd7379-3fe9-45e8-8f69-a8e20fa007a7"}})
; {:headers {"x-request-id" "c9bd7379-3fe9-45e8-8f69-a8e20fa007a7"},
;  :request-id "c9bd7379-3fe9-45e8-8f69-a8e20fa007a7"}
  )

(defn- default-log-fn [{:keys [level msg attrs]}]
  (.println System/out (format "%s | %s -- %s" (s/upper-case (name level)) msg attrs)))

(defn- extract-method-name [request]
  (s/upper-case (name (:request-method request))))

(defn- extract-path [request]
  (str (:uri request)
       (when-let [query-string (:query-string request)]
         (str "?" query-string))))

(defn wrap-request-start-logger
  "Ring middleware that logs information about the start of the request."
  [handler & {:keys [log-fn] :or {log-fn default-log-fn}}]
  (fn [request]
    (let [method (extract-method-name request)
          path (extract-path request)]
      (log-fn {:level :info
               :msg (format "Started %s '%s'" method path)
               :attrs {:method method :path path}})
      (handler (assoc request :start-ms (System/currentTimeMillis))))))

(comment
  (def handler (wrap-request-start-logger identity))

  (handler {:request-method :get
            :uri "/foo"
            :query-string "bar=2"}))

(defn- calculate-duration [{:keys [start-ms]}]
  (if start-ms
    (str (- (System/currentTimeMillis) start-ms) "ms")
    "?ms"))

(defn wrap-request-finish-logger
  "Ring middleware that logs information about the end of the request."
  [handler & {:keys [log-fn] :or {log-fn default-log-fn}}]
  (fn [request]
    (let [method (extract-method-name request)
          path (extract-path request)
          response (handler request)
          status (:status response)
          duration (calculate-duration request)]
      (log-fn {:level :info
               :msg (format "Completed %s '%s' %s in %s" method path status duration)
               :attrs {:method method :path path :status status :duration duration}})
      response)))

(comment
  (def handler (wrap-request-finish-logger #(assoc % :status 200)))

  (handler {:request-method :get
            :start-ms (System/currentTimeMillis)
            :uri "/foo"
            :query-string "bar=2"}))

(def ^:private default-redact-key?
  #{:authorization :password :token :secret :secret-key :secret-token})

(defn wrap-request-params-logger
  "Ring middleware that logs information about the request parameters.

  By default any parameters with key names that are members of the set
  `redact-key?` are automatically filtered. If `redact-key?` is not given
  `default-redact-key?` is used."
  [handler & {:keys [log-fn redact-key?] :or {log-fn default-log-fn
                                              redact-key? default-redact-key?}}]
  (fn [{:keys [params] :as request}]
    (when-not (empty? params)
      (log-fn {:level :info
               :msg "Request parameters"
               :attrs {:params (redact-map (dissoc params :__anti-forgery-token)
                                           {:redact-key? redact-key?
                                            :redact-value "[FILTERED]"})}}))
    (handler request)))

(comment
  (def handler (wrap-request-params-logger identity))

  (handler {:request-method :get
            :start-ms (System/currentTimeMillis)
            :uri "/foo"
            :params {:foo {:bar 42}
                     "password" "secret"}
            :query-string "bar=2"}))
